contracts/0.4.24/Lido.sol

```diff
@@ -651,6 +720,133 @@ contract Lido is ILido, StETH, AragonApp {
+            if (movedToWithdrawalBuffer > 0) {
+                BUFFERED_ETHER_POSITION.setStorageUint256(
+                    _getBufferedEther().add(executionLayerRewards).add(uint256(movedToWithdrawalBuffer))
+                );
+            } else {
+                BUFFERED_ETHER_POSITION.setStorageUint256(
+                    _getBufferedEther().add(executionLayerRewards).sub(uint256(-movedToWithdrawalBuffer))
+                );
+            }
+        }
+    }
+
+    /**
+     * @dev finalize requests in the queue, burn shares and restake some ether if remains
+     * @return withdrawalFundsMovement amount of funds restaked (if positive) or moved to withdrawal buffer (if negative)
+     */
+    function _processWithdrawals(
+        uint256[] _requestIdToFinalizeUpTo,
+        uint256[] _finalizationPooledEtherAmount,
+        uint256[] _finalizationSharesAmount,
+        uint256 _wcBufferedEther
+    ) internal returns (int256 withdrawalFundsMovement) {
+        address withdrawalAddress = address(uint160(getWithdrawalCredentials()));
+        // do nothing if withdrawals is not configured
+        if (withdrawalAddress == address(0)) { 
+            return 0; 
+        }
+
+        IWithdrawalQueue withdrawal = IWithdrawalQueue(withdrawalAddress);
+
+        uint256 lockedEtherAccumulator = 0;
+
+        for (uint256 i = 0; i < _requestIdToFinalizeUpTo.length; i++) {
+            uint256 lastIdToFinalize = _requestIdToFinalizeUpTo[i];
+            require(lastIdToFinalize >= withdrawal.finalizedQueueLength(), "BAD_FINALIZATION_PARAMS");
+
+            uint256 totalPooledEther = _finalizationPooledEtherAmount[i];
+            uint256 totalShares = _finalizationSharesAmount[i];
+
+            (uint256 etherToLock, uint256 sharesToBurn) = withdrawal.calculateFinalizationParams(
+                lastIdToFinalize,
+                totalPooledEther,
+                totalShares 
+            );
+
+            _burnShares(withdrawalAddress, sharesToBurn);
+
+            uint256 remainingFunds = _wcBufferedEther > lockedEtherAccumulator ? _wcBufferedEther - lockedEtherAccumulator: 0;
+
+            uint256 transferToWithdrawalBuffer = etherToLock > remainingFunds ? etherToLock - remainingFunds : 0;
+
+            lockedEtherAccumulator += etherToLock;
+
+            withdrawal.finalize.value(transferToWithdrawalBuffer)(
+                lastIdToFinalize,
+                etherToLock, 
+                totalPooledEther,
+                totalShares
+            );
+        }
+        // There can be unnacounted ether in witdrawal buffer that should not be used for finalization
+        require(lockedEtherAccumulator <= _wcBufferedEther.add(_getBufferedEther()), "NOT_ENOUGH_ACCOUNTED_ETHER");
+
+        withdrawalFundsMovement = int256(_wcBufferedEther) - int256(lockedEtherAccumulator);
+
+        if (withdrawalFundsMovement > 0) {    
+            withdrawal.restake(uint256(withdrawalFundsMovement));
```

Comment:
From my point of view, `deposit` happens when funds transferred to deposit contract. But when used sends their funds to our contract, they `stake`. So, it's appropriate to call it `restake`, because it does not transfer funds to deposit contract.